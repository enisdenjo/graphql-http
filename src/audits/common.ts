/**
 *
 * audit/common
 *
 */

/**
 * Audit requirement levels as per [RFC2119](https://www.rfc-editor.org/rfc/rfc2119).
 */
export type AuditRequirement = 'MUST' | 'SHOULD' | 'MAY';

/**
 * Audit name starting with the audit requirement level.
 */
export type AuditName = `${AuditRequirement} ${string}`;

export interface Audit {
  name: AuditName;
  fn: () => Promise<AuditResult>;
}

export interface AuditOk {
  name: AuditName;
  status: 'ok';
}

export interface AuditFail {
  name: AuditName;
  status: 'warn' | 'error';
  reason: string;
}

export type AuditResult = AuditOk | AuditFail;
