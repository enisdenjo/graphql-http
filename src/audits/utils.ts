/**
 *
 * audit/utils
 *
 */

import type { ExecutionResult } from 'graphql';
import { Audit, AuditName } from './common';

export * from '../utils';

/**
 * Wrap and prepare an audit for testing.
 *
 * @private
 */
export function audit(name: AuditName, fn: () => Promise<void>): Audit {
  return {
    name,
    fn: async () => {
      try {
        await fn();
        return {
          name,
          status: 'ok',
        };
      } catch (errOrReason) {
        if (typeof errOrReason !== 'string') {
          // anything thrown that is not an assertion string is considered fatal
          throw errOrReason;
        }
        return {
          name,
          status: name.startsWith('MUST')
            ? // only failing MUSTs are considered errors
              'error'
            : // everything else is optional and considered a warning
              'warn',
          reason: errOrReason,
        };
      }
    },
  };
}

/**
 * Will throw a string if the assertion fails.
 *
 * All fatal problems will throw an instance of Error.
 *
 * @private
 */
export function assert<T = unknown>(name: string, actual: T) {
  return {
    toBe: (expected: T) => {
      if (actual !== expected) {
        throw `${name} ${actual} is not ${expected}`;
      }
    },
    toBeBetween: (
      min: T extends number ? T : never,
      max: T extends number ? T : never,
    ) => {
      if (!(min <= actual && actual <= max)) {
        throw `${name} ${actual} is not between ${min} and ${max}`;
      }
    },
    toContain: (
      expected: T extends Array<infer U> ? U : T extends string ? T : never,
    ) => {
      // @ts-expect-error types will match, otherwise never
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      if (!actual.includes(expected as any)) {
        throw `${name} ${JSON.stringify(
          actual,
        )} does not contain ${JSON.stringify(expected)}`;
      }
    },
    notToContain: (
      expected: T extends Array<infer U> ? U : T extends string ? T : never,
    ) => {
      // @ts-expect-error types will match, otherwise never
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      if (actual.includes(expected as any)) {
        throw `${name} ${JSON.stringify(actual)} contains ${JSON.stringify(
          expected,
        )}`;
      }
    },
    toHaveProperty: (
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      prop: T extends Record<any, any> ? PropertyKey : never,
    ) => {
      // @ts-expect-error types will match, otherwise never
      if (!(prop in actual)) {
        throw `${name} ${JSON.stringify(
          actual,
        )} does not have a property '${String(prop)}'`;
      }
    },
    notToHaveProperty: (
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      prop: T extends Record<any, any> ? PropertyKey : never,
    ) => {
      // @ts-expect-error types will match, otherwise never
      if (prop in actual) {
        throw `${name} ${JSON.stringify(actual)} does have a property '${String(
          prop,
        )}'`;
      }
    },
  };
}

/**
 * Parses the string as JSON and safely reports parsing issues for audits.
 *
 * Assumes the parsed JSON will be an `ExecutionResult`.
 *
 * @private
 * */
export async function assertBodyAsExecutionResult(res: Response) {
  const str = await res.text();
  try {
    return JSON.parse(str) as ExecutionResult;
  } catch (err) {
    throw `Response body is not valid JSON. Got ${JSON.stringify(str)}`;
  }
}
