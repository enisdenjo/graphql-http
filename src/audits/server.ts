/**
 *
 * audit/server
 *
 */

import type { ExecutionResult } from 'graphql';
import { assert, audit, Audit, AuditResult } from './common';
import { extendedTypeof } from '../utils';

export interface ServerAuditOptions {
  /**
   * The URL of the GraphQL server for the audit.
   */
  url: string;
  /**
   * The Fetch function to use.
   *
   * For NodeJS environments consider using [`@whatwg-node/fetch`](https://github.com/ardatan/whatwg-node/tree/master/packages/fetch).
   *
   * @default global.fetch
   */
  fetchFn?: unknown;
}

export function serverAudits(opts: ServerAuditOptions): Audit[] {
  const fetchFn = (opts.fetchFn || fetch) as typeof fetch;
  return [
    // Media Types
    audit(
      'MUST accept application/graphql-response+json and match the content-type',
      async () => {
        const url = new URL(opts.url);
        url.searchParams.set('query', '{ __typename }');

        const res = await fetchFn(url.toString(), {
          headers: {
            accept: 'application/graphql-response+json',
          },
        });
        assert(res.status).toBe(200);
        assert(res.headers.get('content-type')).toContain(
          'application/graphql-response+json',
        );
      },
    ),
    audit(
      'MUST accept application/json and match the content-type',
      async () => {
        const url = new URL(opts.url);
        url.searchParams.set('query', '{ __typename }');

        const res = await fetchFn(url.toString(), {
          headers: {
            accept: 'application/json',
          },
        });
        assert(res.status).toBe(200);
        assert(res.headers.get('content-type')).toContain('application/json');
      },
    ),
    audit(
      'MUST accept */* and use application/graphql-response+json for the content-type',
      async () => {
        const url = new URL(opts.url);
        url.searchParams.set('query', '{ __typename }');

        const res = await fetchFn(url.toString(), {
          headers: {
            accept: '*/*',
          },
        });
        assert(res.status).toBe(200);
        assert(res.headers.get('content-type')).toContain(
          'application/graphql-response+json',
        );
      },
    ),
    audit(
      'MUST assume application/graphql-response+json content-type when accept is missing',
      async () => {
        const url = new URL(opts.url);
        url.searchParams.set('query', '{ __typename }');

        const res = await fetchFn(url.toString());
        assert(res.status).toBe(200);
        assert(res.headers.get('content-type')).toContain(
          'application/graphql-response+json',
        );
      },
    ),
    audit('MUST use utf-8 charset in response', async () => {
      const url = new URL(opts.url);
      url.searchParams.set('query', '{ __typename }');

      const res = await fetchFn(url.toString());
      assert(res.status).toBe(200);
      assert(res.headers.get('content-type')).toContain('charset=utf-8');
    }),
    audit('MUST accept only utf-8 charset', async () => {
      const url = new URL(opts.url);
      url.searchParams.set('query', '{ __typename }');

      const res = await fetchFn(url.toString(), {
        headers: {
          accept: 'application/graphql-response+json; charset=iso-8859-1',
        },
      });
      assert(res.status).toBe(406);
      assert(res.headers.get('accept')).toContain('charset=utf-8');
    }),
    // Request
    audit('MUST accept POST requests', async () => {
      const res = await fetchFn(opts.url, {
        method: 'POST',
        headers: { 'content-type': 'application/json' },
        body: JSON.stringify({ query: '{ __typename }' }),
      });
      assert(res.status).toBe(200);
    }),
    audit(
      'MAY accept application/x-www-form-urlencoded formatted GET requests',
      async () => {
        const url = new URL(opts.url);
        url.searchParams.set('query', '{ __typename }');

        const res = await fetchFn(url.toString());
        assert(res.status).toBe(200);
      },
    ),
    // Request GET
    audit('MUST NOT allow executing mutations on GET requests', async () => {
      const url = new URL(opts.url);
      url.searchParams.set('query', 'mutation { __typename }');

      const res = await fetchFn(url.toString(), {
        headers: {
          accept: 'application/graphql-response+json',
        },
      });
      assert(res.status).toBe(405);
    }),
    // Request POST
    audit(
      'SHOULD respond with 4xx status code if content-type is not supplied on POST requests',
      async () => {
        const res = await fetchFn(opts.url, {
          method: 'POST',
        });
        assert(res.status).toBeGreaterThanOrEqual(400);
        assert(res.status).toBeLessThanOrEqual(599);
      },
    ),
    audit('MUST accept application/json POST requests', async () => {
      const res = await fetchFn(opts.url, {
        method: 'POST',
        headers: { 'content-type': 'application/json' },
        body: JSON.stringify({ query: '{ __typename }' }),
      });
      assert(res.status).toBe(200);
    }),
    audit('MUST require a request body on POST', async () => {
      const res = await fetchFn(opts.url, {
        method: 'POST',
        headers: { 'content-type': 'application/json' },
      });
      assert(res.status).toBe(400);
    }),
    // Request Parameters
    audit('MUST require the {query} parameter', async () => {
      const res = await fetchFn(opts.url, {
        method: 'POST',
        headers: {
          'content-type': 'application/json',
          accept: 'application/graphql-response+json',
        },
        body: JSON.stringify({ notquery: '{ __typename }' }),
      });
      assert(res.status).toBe(400);
    }),
    ...[{ obj: 'ect' }, 0, false, ['array']].map((invalid) =>
      audit(
        `MUST NOT allow ${extendedTypeof(invalid)} for the {query} parameter`,
        async () => {
          const res = await fetchFn(opts.url, {
            method: 'POST',
            headers: {
              'content-type': 'application/json',
              accept: 'application/graphql-response+json',
            },
            body: JSON.stringify({
              query: invalid,
            }),
          });
          assert(res.status).toBe(400);
        },
      ),
    ),
    audit('MUST accept a string for the {query} parameter', async () => {
      const res = await fetchFn(opts.url, {
        method: 'POST',
        headers: {
          'content-type': 'application/json',
          accept: 'application/graphql-response+json',
        },
        body: JSON.stringify({
          query: '{ __typename }',
        }),
      });
      assert(res.status).toBe(200);
    }),
    ...[{ obj: 'ect' }, 0, false, ['array']].map((invalid) =>
      audit(
        `MUST NOT allow ${extendedTypeof(
          invalid,
        )} for the {operationName} parameter`,
        async () => {
          const res = await fetchFn(opts.url, {
            method: 'POST',
            headers: {
              'content-type': 'application/json',
              accept: 'application/graphql-response+json',
            },
            body: JSON.stringify({
              operationName: invalid,
              query: '{ __typename }',
            }),
          });
          assert(res.status).toBe(400);
        },
      ),
    ),
    audit(
      'MUST accept a string for the {operationName} parameter',
      async () => {
        const res = await fetchFn(opts.url, {
          method: 'POST',
          headers: {
            'content-type': 'application/json',
            accept: 'application/graphql-response+json',
          },
          body: JSON.stringify({
            operationName: 'Query',
            query: 'query Query { __typename }',
          }),
        });
        assert(res.status).toBe(200);
      },
    ),
    ...['string', 0, false, ['array']].map((invalid) =>
      audit(
        `MUST NOT allow ${extendedTypeof(
          invalid,
        )} for the {variables} parameter`,
        async () => {
          const res = await fetchFn(opts.url, {
            method: 'POST',
            headers: {
              'content-type': 'application/json',
              accept: 'application/graphql-response+json',
            },
            body: JSON.stringify({
              query: '{ __typename }',
              variables: invalid,
            }),
          });
          assert(res.status).toBe(400);
        },
      ),
    ),
    audit('MUST accept a map for the {variables} parameter', async () => {
      const res = await fetchFn(opts.url, {
        method: 'POST',
        headers: {
          'content-type': 'application/json',
          accept: 'application/graphql-response+json',
        },
        body: JSON.stringify({
          query: 'query Type($name: String!) { __type(name: $name) { name } }',
          variables: { name: 'sometype' },
        }),
      });
      assert(res.status).toBe(200);
      const result = (await res.json()) as ExecutionResult;
      assert(result).notToHaveProperty('errors');
    }),
    audit(
      'MUST accept a URL-encoded JSON string for the {variables} parameter in GETs',
      async () => {
        const url = new URL(opts.url);
        url.searchParams.set(
          'query',
          'query Type($name: String!) { __type(name: $name) { name } }',
        );
        url.searchParams.set('variables', JSON.stringify({ name: 'sometype' }));
        const res = await fetchFn(url.toString(), {
          method: 'GET',
        });
        assert(res.status).toBe(200);
        const result = (await res.json()) as ExecutionResult;
        assert(result).notToHaveProperty('errors');
      },
    ),
    ...['string', 0, false, ['array']].map((invalid) =>
      audit('MUST NOT allow `%j` for the {extensions} parameter', async () => {
        const res = await fetchFn(opts.url, {
          method: 'POST',
          headers: {
            'content-type': 'application/json',
            accept: 'application/graphql-response+json',
          },
          body: JSON.stringify({
            query: '{ __typename }',
            extensions: invalid,
          }),
        });
        assert(res.status).toBe(400);
      }),
    ),
    audit('MUST accept a map for the {extensions} parameter', async () => {
      const res = await fetchFn(opts.url, {
        method: 'POST',
        headers: {
          'content-type': 'application/json',
          accept: 'application/graphql-response+json',
        },
        body: JSON.stringify({
          query: '{ __typename }',
          extensions: { some: 'value' },
        }),
      });
      assert(res.status).toBe(200);
    }),
    // TODO: audit('MUST accept a map for the {extensions} parameter'),
    // Response application/json
    audit(
      'SHOULD use 200 status code on JSON parsing failure when accepting application/json',
      async () => {
        const res = await fetchFn(opts.url, {
          method: 'POST',
          headers: {
            'content-type': 'application/json',
            accept: 'application/json',
          },
          body: '{ "not a JSON',
        });
        assert(res.status).toBe(200);
      },
    ),
    audit(
      'SHOULD use 200 status code if parameters are invalid when accepting application/json',
      async () => {
        const url = new URL(opts.url);
        url.searchParams.set('qeury' /* typo */, '{ __typename }');
        const res = await fetchFn(url.toString(), {
          method: 'GET',
          headers: { accept: 'application/json' },
        });
        assert(res.status).toBe(200);
      },
    ),
    audit(
      'SHOULD use 200 status code on document parsing failure when accepting application/json',
      async () => {
        const url = new URL(opts.url);
        url.searchParams.set('query', '{');
        const res = await fetchFn(url.toString(), {
          method: 'GET',
          headers: { accept: 'application/json' },
        });
        assert(res.status).toBe(200);
      },
    ),
    audit(
      'SHOULD use 200 status code on document validation failure when accepting application/json',
      async () => {
        const url = new URL(opts.url);
        url.searchParams.set('query', '{ 8f31403dfe404bccbb0e835f2629c6a7 }'); // making sure the field doesnt exist
        const res = await fetchFn(url.toString(), {
          method: 'GET',
          headers: { accept: 'application/json' },
        });
        assert(res.status).toBe(200);
      },
    ),
    // Response application/graphql-response+json
    audit(
      'MUST use 4xx or 5xx status codes on JSON parsing failure when accepting application/graphql-response+json',
      async () => {
        const res = await fetchFn(opts.url, {
          method: 'POST',
          headers: {
            'content-type': 'application/json',
            accept: 'application/graphql-response+json',
          },
          body: '{ "not a JSON',
        });
        assert(res.status).toBeGreaterThanOrEqual(400);
        assert(res.status).toBeLessThanOrEqual(599);
      },
    ),
    audit(
      'SHOULD use 400 status code on JSON parsing failure when accepting application/graphql-response+json',
      async () => {
        const res = await fetchFn(opts.url, {
          method: 'POST',
          headers: {
            'content-type': 'application/json',
            accept: 'application/graphql-response+json',
          },
          body: '{ "not a JSON',
        });
        assert(res.status).toBe(400);
      },
    ),
    audit(
      'MUST use 4xx or 5xx status codes if parameters are invalid when accepting application/graphql-response+json',
      async () => {
        const url = new URL(opts.url);
        url.searchParams.set('qeury' /* typo */, '{ __typename }');
        const res = await fetchFn(url.toString(), {
          method: 'GET',
          headers: { accept: 'application/graphql-response+json' },
        });
        assert(res.status).toBeGreaterThanOrEqual(400);
        assert(res.status).toBeLessThanOrEqual(599);
      },
    ),
    audit(
      'SHOULD use 400 status code if parameters are invalid when accepting application/graphql-response+json',
      async () => {
        const url = new URL(opts.url);
        url.searchParams.set('qeury' /* typo */, '{ __typename }');
        const res = await fetchFn(url.toString(), {
          method: 'GET',
          headers: { accept: 'application/graphql-response+json' },
        });
        assert(res.status).toBe(400);
      },
    ),
    audit(
      'MUST use 4xx or 5xx status codes on document parsing failure when accepting application/graphql-response+json',
      async () => {
        const url = new URL(opts.url);
        url.searchParams.set('query', '{');
        const res = await fetchFn(url.toString(), {
          method: 'GET',
          headers: { accept: 'application/graphql-response+json' },
        });
        assert(res.status).toBeGreaterThanOrEqual(400);
        assert(res.status).toBeLessThanOrEqual(599);
      },
    ),
    audit(
      'SHOULD use 400 status code on document parsing failure when accepting application/graphql-response+json',
      async () => {
        const url = new URL(opts.url);
        url.searchParams.set('query', '{');
        const res = await fetchFn(url.toString(), {
          method: 'GET',
          headers: { accept: 'application/graphql-response+json' },
        });
        assert(res.status).toBe(400);
      },
    ),
    audit(
      'MUST use 4xx or 5xx status codes on document validation failure when accepting application/graphql-response+json',
      async () => {
        const url = new URL(opts.url);
        url.searchParams.set('query', '{ 8f31403dfe404bccbb0e835f2629c6a7 }'); // making sure the field doesnt exist
        const res = await fetchFn(url.toString(), {
          method: 'GET',
          headers: { accept: 'application/graphql-response+json' },
        });
        assert(res.status).toBeGreaterThanOrEqual(400);
        assert(res.status).toBeLessThanOrEqual(599);
      },
    ),
    audit(
      'SHOULD use 400 status code on document validation failure when accepting application/graphql-response+json',
      async () => {
        const url = new URL(opts.url);
        url.searchParams.set('query', '{ 8f31403dfe404bccbb0e835f2629c6a7 }'); // making sure the field doesnt exist
        const res = await fetchFn(url.toString(), {
          method: 'GET',
          headers: { accept: 'application/graphql-response+json' },
        });
        assert(res.status).toBe(400);
      },
    ),
    // TODO: how to make an unauthorized request?
    // https://graphql.github.io/graphql-over-http/draft/#sel-EANNNDTAAEVBAAqqc
    // audit('SHOULD use 401 or 403 status codes when the request is not permitted')
  ];
}

export async function auditServer(
  opts: ServerAuditOptions,
): Promise<AuditResult[]> {
  const audits = serverAudits(opts);

  // audit tests will throw only on fatal errors, tests are contained within the AuditResult
  return await Promise.all(audits.map(({ fn }) => fn()));
}
