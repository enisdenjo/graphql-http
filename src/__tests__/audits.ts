import { Audit, serverAudits } from '../audits';

it('should have globally unique audit ids', () => {
  const ids: string[] = [];

  for (const audit of serverAudits({
    url: 'http://localhost',
    fetchFn: () => {
      // noop
    },
  })) {
    expect(ids).not.toContain(audit.id);
    ids.push(audit.id);
  }
});

it('should not change globally unique audit ids', () => {
  const audits: Omit<Audit, 'fn'>[] = [];

  serverAudits({
    url: 'http://localhost',
    fetchFn: () => {
      // noop
    },
  }).forEach(({ fn, ...audit }) => audits.push(audit));

  // update me if new audits are added or deleted,
  // but existing ones SHOULD NOT CHANGE semantically
  expect(audits).toMatchSnapshot();
});
