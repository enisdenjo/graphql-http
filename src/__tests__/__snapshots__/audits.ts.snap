// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should not change globally unique audit ids 1`] = `
[
  {
    "id": "22EB",
    "name": "SHOULD accept application/graphql-response+json and match the content-type",
  },
  {
    "id": "4655",
    "name": "MUST accept application/json and match the content-type",
  },
  {
    "id": "47DE",
    "name": "SHOULD accept */* and use application/json for the content-type",
  },
  {
    "id": "80D8",
    "name": "SHOULD assume application/json content-type when accept is missing",
  },
  {
    "id": "82A3",
    "name": "MUST use utf-8 encoding when responding",
  },
  {
    "id": "BF61",
    "name": "MUST accept utf-8 encoded request",
  },
  {
    "id": "78D5",
    "name": "MUST assume utf-8 in request if encoding is unspecified",
  },
  {
    "id": "2C94",
    "name": "MUST accept POST requests",
  },
  {
    "id": "5A70",
    "name": "MAY accept application/x-www-form-urlencoded formatted GET requests",
  },
  {
    "id": "9C48",
    "name": "MAY NOT allow executing mutations on GET requests",
  },
  {
    "id": "9ABE",
    "name": "SHOULD respond with 4xx status code if content-type is not supplied on POST requests",
  },
  {
    "id": "03D4",
    "name": "MUST accept application/json POST requests",
  },
  {
    "id": "7267",
    "name": "MUST require a request body on POST",
  },
  {
    "id": "6610",
    "name": "SHOULD use 400 status code on missing {query} parameter when accepting application/graphql-response+json",
  },
  {
    "id": "3715",
    "name": "SHOULD use 200 status code with errors field on missing {query} parameter when accepting application/json",
  },
  {
    "id": "4F50",
    "name": "SHOULD use 400 status code on object {query} parameter when accepting application/graphql-response+json",
  },
  {
    "id": "4F51",
    "name": "SHOULD use 400 status code on number {query} parameter when accepting application/graphql-response+json",
  },
  {
    "id": "4F52",
    "name": "SHOULD use 400 status code on boolean {query} parameter when accepting application/graphql-response+json",
  },
  {
    "id": "4F53",
    "name": "SHOULD use 400 status code on array {query} parameter when accepting application/graphql-response+json",
  },
  {
    "id": "9FE0",
    "name": "SHOULD use 200 status code with errors field on object {query} parameter when accepting application/json",
  },
  {
    "id": "9FE1",
    "name": "SHOULD use 200 status code with errors field on number {query} parameter when accepting application/json",
  },
  {
    "id": "9FE2",
    "name": "SHOULD use 200 status code with errors field on boolean {query} parameter when accepting application/json",
  },
  {
    "id": "9FE3",
    "name": "SHOULD use 200 status code with errors field on array {query} parameter when accepting application/json",
  },
  {
    "id": "34A2",
    "name": "SHOULD allow string {query} parameter when accepting application/graphql-response+json",
  },
  {
    "id": "13EE",
    "name": "MUST allow string {query} parameter when accepting application/json",
  },
  {
    "id": "E3E0",
    "name": "SHOULD use 400 status code on object {operationName} parameter when accepting application/graphql-response+json",
  },
  {
    "id": "E3E1",
    "name": "SHOULD use 400 status code on number {operationName} parameter when accepting application/graphql-response+json",
  },
  {
    "id": "E3E2",
    "name": "SHOULD use 400 status code on boolean {operationName} parameter when accepting application/graphql-response+json",
  },
  {
    "id": "E3E3",
    "name": "SHOULD use 400 status code on array {operationName} parameter when accepting application/graphql-response+json",
  },
  {
    "id": "FB90",
    "name": "SHOULD use 200 status code with errors field on object {operationName} parameter when accepting application/json",
  },
  {
    "id": "FB91",
    "name": "SHOULD use 200 status code with errors field on number {operationName} parameter when accepting application/json",
  },
  {
    "id": "FB92",
    "name": "SHOULD use 200 status code with errors field on boolean {operationName} parameter when accepting application/json",
  },
  {
    "id": "FB93",
    "name": "SHOULD use 200 status code with errors field on array {operationName} parameter when accepting application/json",
  },
  {
    "id": "8161",
    "name": "SHOULD allow string {operationName} parameter when accepting application/graphql-response+json",
  },
  {
    "id": "B8B3",
    "name": "MUST allow string {operationName} parameter when accepting application/json",
  },
  {
    "id": "94B0",
    "name": "SHOULD allow null {variables} parameter when accepting application/graphql-response+json",
  },
  {
    "id": "0220",
    "name": "MUST allow null {variables} parameter when accepting application/json",
  },
  {
    "id": "94B1",
    "name": "SHOULD allow null {operationName} parameter when accepting application/graphql-response+json",
  },
  {
    "id": "0221",
    "name": "MUST allow null {operationName} parameter when accepting application/json",
  },
  {
    "id": "94B2",
    "name": "SHOULD allow null {extensions} parameter when accepting application/graphql-response+json",
  },
  {
    "id": "0222",
    "name": "MUST allow null {extensions} parameter when accepting application/json",
  },
  {
    "id": "69B0",
    "name": "SHOULD use 400 status code on string {variables} parameter when accepting application/graphql-response+json",
  },
  {
    "id": "69B1",
    "name": "SHOULD use 400 status code on number {variables} parameter when accepting application/graphql-response+json",
  },
  {
    "id": "69B2",
    "name": "SHOULD use 400 status code on boolean {variables} parameter when accepting application/graphql-response+json",
  },
  {
    "id": "69B3",
    "name": "SHOULD use 400 status code on array {variables} parameter when accepting application/graphql-response+json",
  },
  {
    "id": "F050",
    "name": "SHOULD use 200 status code with errors field on string {variables} parameter when accepting application/json",
  },
  {
    "id": "F051",
    "name": "SHOULD use 200 status code with errors field on number {variables} parameter when accepting application/json",
  },
  {
    "id": "F052",
    "name": "SHOULD use 200 status code with errors field on boolean {variables} parameter when accepting application/json",
  },
  {
    "id": "F053",
    "name": "SHOULD use 200 status code with errors field on array {variables} parameter when accepting application/json",
  },
  {
    "id": "2EA1",
    "name": "SHOULD allow map {variables} parameter when accepting application/graphql-response+json",
  },
  {
    "id": "28B9",
    "name": "MUST allow map {variables} parameter when accepting application/json",
  },
  {
    "id": "D6D5",
    "name": "MAY allow URL-encoded JSON string {variables} parameter in GETs when accepting application/graphql-response+json",
  },
  {
    "id": "6A70",
    "name": "MAY allow URL-encoded JSON string {variables} parameter in GETs when accepting application/json",
  },
  {
    "id": "9040",
    "name": "SHOULD use 400 status code on string {extensions} parameter when accepting application/graphql-response+json",
  },
  {
    "id": "9041",
    "name": "SHOULD use 400 status code on number {extensions} parameter when accepting application/graphql-response+json",
  },
  {
    "id": "9042",
    "name": "SHOULD use 400 status code on boolean {extensions} parameter when accepting application/graphql-response+json",
  },
  {
    "id": "9043",
    "name": "SHOULD use 400 status code on array {extensions} parameter when accepting application/graphql-response+json",
  },
  {
    "id": "3680",
    "name": "SHOULD use 200 status code with errors field on string {extensions} parameter when accepting application/json",
  },
  {
    "id": "3681",
    "name": "SHOULD use 200 status code with errors field on number {extensions} parameter when accepting application/json",
  },
  {
    "id": "3682",
    "name": "SHOULD use 200 status code with errors field on boolean {extensions} parameter when accepting application/json",
  },
  {
    "id": "3683",
    "name": "SHOULD use 200 status code with errors field on array {extensions} parameter when accepting application/json",
  },
  {
    "id": "428F",
    "name": "SHOULD allow map {extensions} parameter when accepting application/graphql-response+json",
  },
  {
    "id": "1B7A",
    "name": "MUST allow map {extensions} parameter when accepting application/json",
  },
  {
    "id": "D477",
    "name": "SHOULD use 200 status code on JSON parsing failure when accepting application/json",
  },
  {
    "id": "F5AF",
    "name": "SHOULD use 200 status code if parameters are invalid when accepting application/json",
  },
  {
    "id": "572B",
    "name": "SHOULD use 200 status code on document parsing failure when accepting application/json",
  },
  {
    "id": "FDE2",
    "name": "SHOULD use 200 status code on document validation failure when accepting application/json",
  },
  {
    "id": "60AA",
    "name": "SHOULD use 4xx or 5xx status codes on JSON parsing failure when accepting application/graphql-response+json",
  },
  {
    "id": "2163",
    "name": "SHOULD use 400 status code on JSON parsing failure when accepting application/graphql-response+json",
  },
  {
    "id": "3E36",
    "name": "SHOULD use 4xx or 5xx status codes if parameters are invalid when accepting application/graphql-response+json",
  },
  {
    "id": "17C5",
    "name": "SHOULD use 400 status code if parameters are invalid when accepting application/graphql-response+json",
  },
  {
    "id": "34D6",
    "name": "SHOULD not contain the data entry if parameters are invalid when accepting application/graphql-response+json",
  },
  {
    "id": "865D",
    "name": "SHOULD use 4xx or 5xx status codes on document parsing failure when accepting application/graphql-response+json",
  },
  {
    "id": "556A",
    "name": "SHOULD use 400 status code on document parsing failure when accepting application/graphql-response+json",
  },
  {
    "id": "D586",
    "name": "SHOULD not contain the data entry on document parsing failure when accepting application/graphql-response+json",
  },
  {
    "id": "51FE",
    "name": "SHOULD use 4xx or 5xx status codes on document validation failure when accepting application/graphql-response+json",
  },
  {
    "id": "74FF",
    "name": "SHOULD use 400 status code on document validation failure when accepting application/graphql-response+json",
  },
  {
    "id": "5E5B",
    "name": "SHOULD not contain the data entry on document validation failure when accepting application/graphql-response+json",
  },
  {
    "id": "86EE",
    "name": "SHOULD use a status code of 400 on variable coercion failure when accepting application/graphql-response+json",
  },
]
`;
